In [1]:
import pandas as pd

    В этом руководстве используется набор данных Titanic, сохраненный в формате CSV. Данные состоят из следующих столбцов данных:
    
    * PassengerId: идентификатор каждого пассажира.
    * Survived: указывает, выжил ли пассажир. `0` за да и `1` за нет.
    * Pclass: один из 3 классов билетов: Class `1`, Class `2` и `Class 3`.
    * Name: Имя пассажира.
    * Sex: Пол пассажира.
    * Age: Возраст пассажира в годах.
    * SibSp: Количество братьев и сестер или супругов на борту.
    * Parch: Количество родителей или детей на борту.
    * Ticket: номер билета пассажира.
    * Fare: указание тарифа.
    * Cabin: Номер кабины пассажира.
    * Embarked: Порт посадки.

# Как читать и записывать табличные данные?

![02_io_readwrite](https://pandas.pydata.org/docs/_images/02_io_readwrite.svg)

___

⸮ Я хочу проанализировать данные о пассажирах Титаника, доступные в виде CSV-файла.

In [2]:
titanic = pd.read_csv("data/titanic.csv")

pandas предоставляет функцию `read_csv()` для чтения данных, хранящихся в виде файла csv, в `DataFrame`. pandas поддерживает множество различных форматов файлов или источников данных из коробки (csv, excel, sql, json, parquet, …), каждый из них имеет префикс `read_*`.

Всегда проверяйте данные после чтения данных. При отображении `DataFrame` по умолчанию будут отображаться первые и последние 5 строк:

In [3]:
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

___

⸮ Я хочу увидеть первые 8 строк `DataFrame`.

In [4]:
titanic.head(8)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S

*Примечание:* Вместо этого вас интересуют последние N строк? pandas также предоставляет метод `tail()`. Например, `titanic.tail(10)` вернет последние 10 строк DataFrame.

In [5]:
titanic.tail(10)

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

Проверить, как pandas интерпретирует каждый из типов данных столбца, можно, запросив атрибут pandas `dtypes`.

In [6]:
titanic.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

Для каждого столбца указывается используемый тип данных. Типы данных в этом `DataFrame` — целые числа (`int64`), числа с плавающей запятой (`float64`) и строки (`object`).

*Замечание:* При запросе `dtypes` скобки не используются! `dtypes` — это атрибут `DataFrame` и `Series`, а не метод.

___

⸮ Мой коллега запросил данные Титаника в виде электронной таблицы.

In [7]:
titanic.to_excel("data/titanic.xlsx", sheet_name="passengers", index=False)

В то время как функции `read_*` используются для чтения данных в pandas, методы `to_*` используются для хранения данных. Метод `to_excel()` сохраняет данные в виде файла Excel. В приведенном здесь примере `sheet_name` называется `passengers`, а не листом по умолчанию `Sheet1`. При установке `index=False` метки индексов строк не сохраняются в электронной таблице.

Эквивалентная функция чтения `read_excel()` перезагрузит данные в `DataFrame`:

In [8]:
titanic = pd.read_excel("data/titanic.xlsx", sheet_name="passengers")
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

___

⸮ Меня интересует техническое резюме `DataFrame`

In [9]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Метод `info()` предоставляет техническую информацию о `DataFrame`, поэтому давайте объясним вывод более подробно:

* Это действительно `DataFrame`.
* Имеется 891 запись, т.е. 891 строка.
* Каждая строка имеет метку строки (она же `index`) со значениями в диапазоне от 0 до 890.
* В таблице 12 столбцов. Большинство столбцов имеют значение для каждой из строк (все значения 891 `non-null`). В некоторых столбцах отсутствуют значения и менее 891 ненулевого значения.
* Столбцы `Name`, `Sex`, `Cabin` и `Embarked` состоят из текстовых данных (строки, также известные как `object`). Другие столбцы представляют собой числовые данные, причем некоторые из них — целые числа (так называемые `integer`), а другие — действительные числа (так называемые `float`).
* Типы данных (символы, целые числа и т. д.) в разных столбцах суммируются путем перечисления `dtypes`.
* Также указывается приблизительный объем оперативной памяти, используемой для хранения `DataFrame`.